File: 0678_valid_parenthesis/check_valid_string.go
package main

/**
 * Time: O(n) 
 * Space: O(1)
 */
func checkValidString(s string) bool {
	var curr_min, curr_max int
	for i := 0; i < len(s); i++ {
		if s[i] == '(' {
			curr_min++
			curr_max+++
		} else if s[i] == ')' {
			curr_max--
			curr_min = max(curr_min-1, 0)
		} else {
			curr_max++
			curr_min = max(curr_min-1, 0)
		}
		if curr_max < 0 {
			return false
		}
	}
	return curr_min == 0
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {

}

File: 0678_valid_parenthesis/check_valid_string.cpp
#include <iostream>
#include <string>

using namespace std;

/**
 * Time: O(n) 
 * Space: O(1)
 */
bool check_valid_string(string s)
{
    int curr_min = 0, curr_max = 0;
    for (auto i : s)
    {
        curr_min = i == '(' ? curr_min + 1 : max(curr_min - 1, 0);
        curr_max = i == ')' ? curr_max - 1 : curr_max + 1;
        if (curr_max < 0)
            return false;
    }
    return !curr_min;
}
